File: Part8/C++/MazeWithFunctions/MazeWithFunctions.cpp
#include <iostream>
#include <cstdlib>
#include <ctime>
#include <cstring>



// Define directions
const int LEFT = 0;
const int UP = 1;
const int RIGHT = 2;
const int DOWN = 3;

const int FIRST_CELL = 0;
const int SECOND_CELL = 1;

const int NO_CELL = -1;



// Global variables
bool* AllHorizontalWallsUp_Glob;
bool* WallsUp_Glob;
int InteriorWallCount_Glob;
int MazeRows_Glob;
int MazeColumns_Glob;



// +-+-+-+
// Prints a horizontal wall, similar to the example above, 
// with walls down based on the provided parameters.
// HorizontalWallsUp_Par: Boolean array of size mazeColumns, 
//     True indicates the wall should be printed as up.
void printHorizontalWalls ( const bool* HorizontalWallsUp_Par )
{
    printf ( "+" );
    for ( int horizontalWallIndex = 0; horizontalWallIndex < MazeColumns_Glob; horizontalWallIndex++ )
    {
        if ( HorizontalWallsUp_Par[horizontalWallIndex] == true )
        {
            printf ( "-" );
        }
        else
        {
            printf ( " " );
        }
        printf ( "+" );
    }
    printf ( "\n" );
}



// +-+-+-+
// Prints a horizontal wall, similar to the example above, with all walls up.
void printHorizontalWalls()
{
    printHorizontalWalls ( AllHorizontalWallsUp_Glob );
}



// | | | |
// Prints a vertical wall, similar to the example above, 
// with walls down based on the provided parameters.
// VerticalWallsUp_Par: Boolean array of size mazeColumns - 1, 
//     True indicates the wall should be printed as up.
void printVerticalWalls ( const bool* VerticalWallsUp_Par )
{
    // First wall is an exterior wall, always up.
    printf ( "|" );
    for ( int verticalWallIndex = 0; verticalWallIndex < MazeColumns_Glob - 1; verticalWallIndex++ )
    {
        printf ( " " );
        if ( VerticalWallsUp_Par[verticalWallIndex] == true )
        {
            printf ( "|" );
        }
        else
        {
            printf ( " " );
        }
    }
    // Last wall exterior, always up.
    printf ( " " );
    printf ( "|" );
    printf ( "\n" );
}



// Loop through the rows of the maze and print the maze 
// based on WallsUp_Glob
void printMaze()
{
    int interiorWallIndex = 0;

    // First row is exterior walls
    printHorizontalWalls();
    for ( int rowIndex = 0; rowIndex < MazeRows_Glob; rowIndex++ )
    {
        bool verticalWallsUp[MazeColumns_Glob - 1];
        for ( int columnIndex = 0; columnIndex < MazeColumns_Glob - 1; columnIndex++ )
        {
            verticalWallsUp[columnIndex] = WallsUp_Glob[interiorWallIndex];
            interiorWallIndex = interiorWallIndex + 1;
        }

        printVerticalWalls ( verticalWallsUp );

        if ( rowIndex == MazeRows_Glob - 1 )
        {
            printHorizontalWalls();
        }
        else
        {
            bool horizontalWallsUp[MazeColumns_Glob];
            for ( int columnIndex = 0; columnIndex < MazeColumns_Glob; columnIndex++ )
            {
                horizontalWallsUp[columnIndex] = WallsUp_Glob[interiorWallIndex];
                interiorWallIndex = interiorWallIndex + 1;
            }

            printHorizontalWalls ( horizontalWallsUp );
        }
    }

    printf ( "\n" );
}



// Simple sleep function using busy wait
void sleepHalfSecond()
{
    clock_t start = clock();
    while ( ( clock() - start ) < CLOCKS_PER_SEC / 2 );
}



// Kruskal's algorithm.
// The simple description of the algorithm is first place each 
// cell in its own group.  Then process all walls in random order,
// if the cells on either side of the wall are in separate groups, 
// remove the wall and merge the groups.  Repeat until all 
// cells are now in the same group.
void buildMazeKruskal()
{
    // Identify the cells each wall connects.
    int wallConnections[InteriorWallCount_Glob][2];
    int wallRemoveList[InteriorWallCount_Glob];
    int cellToGroup[MazeRows_Glob * MazeColumns_Glob];
    int groupCells[MazeRows_Glob * MazeColumns_Glob][MazeRows_Glob * MazeColumns_Glob];

    int wallIndex = 0;
    for ( int rowIndex = 0; rowIndex < MazeRows_Glob; rowIndex++ )
    {
        // Track the first cell in the current row
        int firstCellInRow = rowIndex * MazeColumns_Glob;

        // Note the 0..mazeColumns minus 2, one less vertical wall
        // than the number of columns.
        for ( int verticalWallIndex = 0; verticalWallIndex < MazeColumns_Glob - 1; verticalWallIndex++ )
        {
            int leftCell = firstCellInRow + verticalWallIndex;
            int rightCell = leftCell + 1;
            wallConnections[wallIndex][FIRST_CELL] = leftCell;
            wallConnections[wallIndex][SECOND_CELL] = rightCell;
            wallIndex = wallIndex + 1;
        }

        // The last row will have no interior horizontal walls below
        // it, so will be skipped.
        if ( wallIndex < InteriorWallCount_Glob )
        {
            for ( int horizontalWallIndex = 0; horizontalWallIndex < MazeColumns_Glob; horizontalWallIndex++ )
            {
                int upperCell = firstCellInRow + horizontalWallIndex;
                int lowerCell = upperCell + MazeColumns_Glob;
                wallConnections[wallIndex][FIRST_CELL] = upperCell;
                wallConnections[wallIndex][SECOND_CELL] = lowerCell;
                wallIndex = wallIndex + 1;
            }
        }
    }

    for ( int cellIndex = 0; cellIndex < MazeRows_Glob * MazeColumns_Glob; cellIndex++ )
    {
        cellToGroup[cellIndex] = cellIndex;

        for ( int innerCellIndex = 0; innerCellIndex < MazeRows_Glob * MazeColumns_Glob; innerCellIndex++ )
        {
            if ( innerCellIndex == 0 )
            {
                groupCells[cellIndex][innerCellIndex] = cellIndex;
            }
            else
            {
                groupCells[cellIndex][innerCellIndex] = NO_CELL;
            }
        }
    }

    bool mazeComplete = false;

    for ( int wallIndex = 0; wallIndex < InteriorWallCount_Glob; wallIndex++ )
    {
        wallRemoveList[wallIndex] = wallIndex;
    }
    
    // Fisher-Yates shuffle
    for ( int i = InteriorWallCount_Glob - 1; i > 0; i-- )
    {
        int j = rand() % ( i + 1 );
        int temp = wallRemoveList[i];
        wallRemoveList[i] = wallRemoveList[j];
        wallRemoveList[j] = temp;
    }

    // Perform Kruskal's algorithm.
    for ( int removeWallIndex = 0; removeWallIndex < InteriorWallCount_Glob; removeWallIndex++ )
    {
        int nextWallToCheck = wallRemoveList[removeWallIndex];

        // If the two cells connected to this wall are not part 
        // of the same group, remove the wall and merge the 
        // groups.
        int firstCell = wallConnections[nextWallToCheck][FIRST_CELL];
        int firstCellGroupIndex = cellToGroup[firstCell];
        int secondCell = wallConnections[nextWallToCheck][SECOND_CELL];
        int secondCellGroupIndex = cellToGroup[secondCell];
        if ( firstCellGroupIndex != secondCellGroupIndex )
        {
            WallsUp_Glob[nextWallToCheck] = false;

            // Loop through the indices of all cells in the first 
            // group until we find a NO_CELL indicating no cell here.
            int nextEmptyFirstGroupIndex = 0;
            for ( int cellIndex = 0; cellIndex < MazeColumns_Glob * MazeRows_Glob; cellIndex++ )
            {
                if ( groupCells[firstCellGroupIndex][cellIndex] == NO_CELL )
                {
                    nextEmptyFirstGroupIndex = cellIndex;
                    break;
                }
            }

            // Loop through the indices of all cells in the second group,
            // move each cell to the first group, and set that cell's 
            // group to the first group index.
            for ( int groupCellIndex = MazeColumns_Glob * MazeRows_Glob - 1; groupCellIndex >= 0; groupCellIndex-- )
            {
                // Skip until we reach valid walls
                if ( groupCells[secondCellGroupIndex][groupCellIndex] != NO_CELL )
                {
                    // Get the id number of the cell to move from 
                    // the second group to the first group
                    int cellToMove = groupCells[secondCellGroupIndex][groupCellIndex];

                    // Move the cell number from the second group 
                    // to the first group
                    groupCells[firstCellGroupIndex][nextEmptyFirstGroupIndex] = cellToMove;
                    // Move our empty index to the next cell in this array.
                    nextEmptyFirstGroupIndex = nextEmptyFirstGroupIndex + 1;
                    // Mark this cell as part of the first group.
                    cellToGroup[cellToMove] = firstCellGroupIndex;
                    // Remove the cell from the second group (set the
                    // array entry to NO_CELL)
                    groupCells[secondCellGroupIndex][groupCellIndex] = NO_CELL;

                    if ( nextEmptyFirstGroupIndex >= MazeColumns_Glob * MazeRows_Glob )
                    {
                        mazeComplete = true;
                    }
                }
            }

            sleepHalfSecond();

            printMaze();

            if ( mazeComplete == true )
            {
                break;
            }
        }
    }
}



// Note we could use command line arguments instead of user 
// prompts to get the maze size.
int main()
{
    // Seed random number generator
    srand ( time ( nullptr ) );

    // Prompt the user for maze size
    MazeColumns_Glob = 0;
    while ( MazeColumns_Glob <= 0 )
    {
        printf ( "Please enter number of columns for maze, must be greater than 1: " );
        char userInput[100];
        if ( fgets ( userInput, sizeof ( userInput ), stdin ) )
        {
            // Remove newline if present
            int len = strlen ( userInput );
            if ( len > 0 && userInput[len - 1] == '\n' )
            {
                userInput[len - 1] = '\0';
            }
            
            // Check if string is a valid number
            bool isValid = true;
            if ( userInput[0] == '\0' )
            {
                isValid = false;
            }
            else
            {
                for ( int i = 0; userInput[i] != '\0'; i++ )
                {
                    if ( userInput[i] < '0' || userInput[i] > '9' )
                    {
                        isValid = false;
                        break;
                    }
                }
            }
            
            if ( isValid == true )
            {
                MazeColumns_Glob = atoi ( userInput );
            }
        }
    }

    MazeRows_Glob = 0;
    while ( MazeRows_Glob <= 0 )
    {
        printf ( "Please enter number of rows for maze, must be greater than 1: " );
        char userInput[100];
        if ( fgets ( userInput, sizeof ( userInput ), stdin ) )
        {
            // Remove newline if present
            int len = strlen ( userInput );
            if ( len > 0 && userInput[len - 1] == '\n' )
            {
                userInput[len - 1] = '\0';
            }
            
            // Check if string is a valid number
            bool isValid = true;
            if ( userInput[0] == '\0' )
            {
                isValid = false;
            }
            else
            {
                for ( int i = 0; userInput[i] != '\0'; i++ )
                {
                    if ( userInput[i] < '0' || userInput[i] > '9' )
                    {
                        isValid = false;
                        break;
                    }
                }
            }
            
            if ( isValid == true )
            {
                MazeRows_Glob = atoi ( userInput );
            }
        }
    }

    // Setup maze datastructures for the user entered size.
    AllHorizontalWallsUp_Glob = new bool[MazeColumns_Glob];
    for ( int i = 0; i < MazeColumns_Glob; i++ )
    {
        AllHorizontalWallsUp_Glob[i] = true;
    }
    
    InteriorWallCount_Glob = MazeRows_Glob * ( MazeColumns_Glob - 1 ) + ( MazeRows_Glob - 1 ) * MazeColumns_Glob;
    
    WallsUp_Glob = new bool[InteriorWallCount_Glob];
    for ( int i = 0; i < InteriorWallCount_Glob; i++ )
    {
        WallsUp_Glob[i] = true;
    }

    buildMazeKruskal();

    // Clean up global memory
    delete[] AllHorizontalWallsUp_Glob;
    delete[] WallsUp_Glob;

    return 0;
}

File: Part8/Fortran/MazeWithFunctions/MazeWithFunctions.f90
program maze_generator
    implicit none
    
    ! Define directions
    integer, parameter :: LEFT = 0
    integer, parameter :: UP = 1
    integer, parameter :: RIGHT = 2
    integer, parameter :: DOWN = 3
    
    integer, parameter :: FIRST_CELL = 1
    integer, parameter :: SECOND_CELL = 2
    
    integer, parameter :: NO_CELL = -1
    
    ! Global variables - now allocatable
    logical, allocatable :: all_horizontal_walls_up_glob(:)
    logical, allocatable :: walls_up_glob(:)
    integer :: interior_wall_count_glob
    integer :: maze_rows_glob
    integer :: maze_columns_glob
    
    call main_program()
    
contains



    ! +-+-+-+
    ! Prints a horizontal wall, similar to the example above, 
    ! with walls down based on the provided parameters.
    subroutine print_horizontal_walls ( horizontal_walls_up_par )
        implicit none
        logical, intent(in) :: horizontal_walls_up_par(:)
        integer :: horizontal_wall_index
        
        write(*,'(A)', advance='no') '+'
        do horizontal_wall_index = 1, maze_columns_glob
            if ( horizontal_walls_up_par(horizontal_wall_index) .eqv. .true. ) then
                write(*,'(A)', advance='no') '-'
            else
                write(*,'(A)', advance='no') ' '
            end if
            write(*,'(A)', advance='no') '+'
        end do
        write(*,*)
    end subroutine print_horizontal_walls



    ! +-+-+-+
    ! Prints a horizontal wall, similar to the example above, with all walls up.
    subroutine print_horizontal_walls_all()
        implicit none
        call print_horizontal_walls ( all_horizontal_walls_up_glob )
    end subroutine print_horizontal_walls_all



    ! | | | |
    ! Prints a vertical wall, similar to the example above, 
    ! with walls down based on the provided parameters.
    subroutine print_vertical_walls ( vertical_walls_up_par )
        implicit none
        logical, intent(in) :: vertical_walls_up_par(:)
        integer :: vertical_wall_index
        
        ! First wall is an exterior wall, always up.
        write(*,'(A)', advance='no') '|'
        do vertical_wall_index = 1, maze_columns_glob - 1
            write(*,'(A)', advance='no') ' '
            if ( vertical_walls_up_par(vertical_wall_index) .eqv. .true. ) then
                write(*,'(A)', advance='no') '|'
            else
                write(*,'(A)', advance='no') ' '
            end if
        end do
        ! Last wall exterior, always up.
        write(*,'(A)', advance='no') ' '
        write(*,'(A)', advance='no') '|'
        write(*,*)
    end subroutine print_vertical_walls



    ! Loop through the rows of the maze and print the maze 
    ! based on walls_up_glob
    subroutine print_maze()
        implicit none
        integer :: interior_wall_index
        integer :: row_index
        integer :: column_index
        logical, allocatable :: vertical_walls_up(:)
        logical, allocatable :: horizontal_walls_up(:)
        
        interior_wall_index = 1
        
        ! First row is exterior walls
        call print_horizontal_walls_all()
        do row_index = 1, maze_rows_glob
            allocate(vertical_walls_up(maze_columns_glob - 1))
            do column_index = 1, maze_columns_glob - 1
                vertical_walls_up(column_index) = walls_up_glob(interior_wall_index)
                interior_wall_index = interior_wall_index + 1
            end do
            
            call print_vertical_walls ( vertical_walls_up )
            deallocate(vertical_walls_up)
            
            if ( row_index .eq. maze_rows_glob ) then
                call print_horizontal_walls_all()
            else
                allocate(horizontal_walls_up(maze_columns_glob))
                do column_index = 1, maze_columns_glob
                    horizontal_walls_up(column_index) = walls_up_glob(interior_wall_index)
                    interior_wall_index = interior_wall_index + 1
                end do
                
                call print_horizontal_walls ( horizontal_walls_up )
                deallocate(horizontal_walls_up)
            end if
        end do
        
        write(*,*)
    end subroutine print_maze



    ! Simple sleep function
    subroutine sleep_half_second()
        implicit none
        ! Use standard sleep (1 second since half-second not available)
        call sleep(1)
    end subroutine sleep_half_second



    ! Kruskal's algorithm.
    subroutine build_maze_kruskal()
        implicit none
        integer, allocatable :: wall_connections(:,:)
        integer, allocatable :: wall_remove_list(:)
        integer, allocatable :: cell_to_group(:)
        integer, allocatable :: group_cells(:,:)
        integer :: wall_index
        integer :: row_index
        integer :: first_cell_in_row
        integer :: vertical_wall_index
        integer :: horizontal_wall_index
        integer :: left_cell, right_cell
        integer :: upper_cell, lower_cell
        integer :: cell_index
        integer :: inner_cell_index
        logical :: maze_complete
        integer :: wall_index_loop
        integer :: i, j, temp
        real :: random_val
        integer :: remove_wall_index
        integer :: next_wall_to_check
        integer :: first_cell_index
        integer :: first_cell_group_index
        integer :: second_cell_index
        integer :: second_cell_group_index
        integer :: next_empty_first_group_index
        integer :: group_cell_index
        integer :: cell_to_move
        integer :: total_cells
        integer :: iostat_val
        
        total_cells = maze_rows_glob * maze_columns_glob
        
        ! Allocate all arrays with exact sizes needed
        allocate(wall_connections(interior_wall_count_glob, 2), stat=iostat_val)
        if (iostat_val /= 0) stop 'Error allocating wall_connections'
        
        allocate(wall_remove_list(interior_wall_count_glob), stat=iostat_val)
        if (iostat_val /= 0) stop 'Error allocating wall_remove_list'
        
        allocate(cell_to_group(total_cells), stat=iostat_val)
        if (iostat_val /= 0) stop 'Error allocating cell_to_group'
        
        allocate(group_cells(total_cells, total_cells), stat=iostat_val)
        if (iostat_val /= 0) stop 'Error allocating group_cells'
        
        wall_index = 1
        do row_index = 1, maze_rows_glob
            ! Track the first cell in the current row
            first_cell_in_row = (row_index - 1) * maze_columns_glob + 1
            
            ! Note the 1..maze_columns_glob - 1, one less vertical wall
            ! than the number of columns.
            do vertical_wall_index = 1, maze_columns_glob - 1
                left_cell = first_cell_in_row + vertical_wall_index - 1
                right_cell = left_cell + 1
                wall_connections(wall_index, FIRST_CELL) = left_cell
                wall_connections(wall_index, SECOND_CELL) = right_cell
                wall_index = wall_index + 1
            end do
            
            ! The last row will have no interior horizontal walls below
            ! it, so will be skipped.
            if ( wall_index .le. interior_wall_count_glob ) then
                do horizontal_wall_index = 1, maze_columns_glob
                    upper_cell = first_cell_in_row + horizontal_wall_index - 1
                    lower_cell = upper_cell + maze_columns_glob
                    wall_connections(wall_index, FIRST_CELL) = upper_cell
                    wall_connections(wall_index, SECOND_CELL) = lower_cell
                    wall_index = wall_index + 1
                end do
            end if
        end do
        
        do cell_index = 1, total_cells
            cell_to_group(cell_index) = cell_index
            
            do inner_cell_index = 1, total_cells
                if ( inner_cell_index .eq. 1 ) then
                    group_cells(cell_index, inner_cell_index) = cell_index
                else
                    group_cells(cell_index, inner_cell_index) = NO_CELL
                end if
            end do
        end do
        
        maze_complete = .false.
        
        do wall_index_loop = 1, interior_wall_count_glob
            wall_remove_list(wall_index_loop) = wall_index_loop
        end do
        
        ! Fisher-Yates shuffle
        do i = interior_wall_count_glob, 2, -1
            call random_number(random_val)
            j = int(random_val * real(i)) + 1
            if ( j .ge. 1 .and. j .le. i ) then
                temp = wall_remove_list(i)
                wall_remove_list(i) = wall_remove_list(j)
                wall_remove_list(j) = temp
            end if
        end do
        
        ! Perform Kruskal's algorithm.
        do remove_wall_index = 1, interior_wall_count_glob
            next_wall_to_check = wall_remove_list(remove_wall_index)
            
            ! If the two cells connected to this wall are not part 
            ! of the same group, remove the wall and merge the 
            ! groups.
            first_cell_index = wall_connections(next_wall_to_check, FIRST_CELL)
            first_cell_group_index = cell_to_group(first_cell_index)
            second_cell_index = wall_connections(next_wall_to_check, SECOND_CELL)
            second_cell_group_index = cell_to_group(second_cell_index)
            if ( first_cell_group_index .ne. second_cell_group_index ) then
                walls_up_glob(next_wall_to_check) = .false.
                
                ! Loop through the indices of all cells in the first 
                ! group until we find a NO_CELL indicating no cell here.
                next_empty_first_group_index = 1
                do cell_index = 1, total_cells
                    if ( group_cells(first_cell_group_index, cell_index) .eq. NO_CELL ) then
                        next_empty_first_group_index = cell_index
                        exit
                    end if
                end do
                
                ! Loop through the indices of all cells in the second group,
                ! move each cell to the first group, and set that cell's 
                ! group to the first group index.
                do group_cell_index = total_cells, 1, -1
                    ! Skip until we reach valid cells
                    if ( group_cells(second_cell_group_index, group_cell_index) .ne. NO_CELL ) then
                        ! Get the id number of the cell to move from 
                        ! the second group to the first group
                        cell_to_move = group_cells(second_cell_group_index, group_cell_index)
                        
                        ! Move the cell number from the second group 
                        ! to the first group
                        group_cells(first_cell_group_index, next_empty_first_group_index) = cell_to_move
                        ! Move our empty index to the next cell in this array.
                        next_empty_first_group_index = next_empty_first_group_index + 1
                        ! Mark this cell as part of the first group.
                        cell_to_group(cell_to_move) = first_cell_group_index
                        ! Remove the cell from the second group (set the
                        ! array entry to NO_CELL)
                        group_cells(second_cell_group_index, group_cell_index) = NO_CELL
                        
                        if ( next_empty_first_group_index .ge. total_cells ) then
                            maze_complete = .true.
                        end if
                    end if
                end do
                
                call sleep_half_second()
                
                call print_maze()
                
                if ( maze_complete .eqv. .true. ) then
                    exit
                end if
            end if
        end do
        
        ! Clean up allocated arrays
        deallocate(wall_connections)
        deallocate(wall_remove_list)
        deallocate(cell_to_group)
        deallocate(group_cells)
    end subroutine build_maze_kruskal



    subroutine main_program()
        implicit none
        character(len=20) :: user_input
        integer :: temp_value
        integer :: ios
        integer :: i
        integer :: iostat_val
        
        ! Initialize random seed
        call random_seed()
        
        ! Prompt the user for maze size
        maze_columns_glob = 0
        do while ( maze_columns_glob .le. 0 )
            write(*,'(A)', advance='no') 'Please enter number of columns for maze, must be greater than 1: '
            read(*,'(A)') user_input
            read(user_input, *, iostat=ios) temp_value
            if ( ios .eq. 0 ) then
                if ( temp_value .gt. 1 ) then
                    maze_columns_glob = temp_value
                end if
            end if
        end do
        
        maze_rows_glob = 0
        do while ( maze_rows_glob .le. 0 )
            write(*,'(A)', advance='no') 'Please enter number of rows for maze, must be greater than 1: '
            read(*,'(A)') user_input
            read(user_input, *, iostat=ios) temp_value
            if ( ios .eq. 0 ) then
                if ( temp_value .gt. 1 ) then
                    maze_rows_glob = temp_value
                end if
            end if
        end do
        
        ! Calculate interior wall count
        interior_wall_count_glob = maze_rows_glob * ( maze_columns_glob - 1 ) + ( maze_rows_glob - 1 ) * maze_columns_glob
        
        ! Allocate global arrays with exact sizes needed
        allocate(all_horizontal_walls_up_glob(maze_columns_glob), stat=iostat_val)
        if (iostat_val /= 0) stop 'Error allocating all_horizontal_walls_up_glob'
        
        allocate(walls_up_glob(interior_wall_count_glob), stat=iostat_val)
        if (iostat_val /= 0) stop 'Error allocating walls_up_glob'
        
        ! Setup maze datastructures for the user entered size.
        do i = 1, maze_columns_glob
            all_horizontal_walls_up_glob(i) = .true.
        end do
        
        do i = 1, interior_wall_count_glob
            walls_up_glob(i) = .true.
        end do
        
        call build_maze_kruskal()
        
        ! Clean up global arrays
        deallocate(all_horizontal_walls_up_glob)
        deallocate(walls_up_glob)
    end subroutine main_program

end program maze_generator
